### DISCUSSION TITLE
Use architecture-native images and binaries (arm64 for mac M1)

### DISCUSSION THREAD

**Discussion by rfay** [lima-vm/lima#406] [General]
I was caught by surprise to see qemu running things on my arm64 M1 mac. 

qemu is a wonderful concoction, but it's not something you want to rely on for real work. It has tremendous problems with reliabilities/crashes and also with performance.

But there's not really a reason I understand not to use native binaries and images. The project is written in golang, it's easy to compile to multiple target architectures, and it's easy to build native docker images. 

Is there already work going on in this area?

**Comment by jandubois** [comment]
> But there's not really a reason I understand not to use native binaries and images.

I don't know what you mean by "native binaries". The goal is to run a container runtime, like `containerd`, and so far they only run on Linux, so on macOS you need a virtual machine to provide the Linux environment. It doesn't have to be qemu, but you need some kind of hypervisor, and given that Lima is an open source project, the hypervisor should also be freely available under an open source license as well (which rules out Parallels, VMware Fusion, etc).

Given that qemu uses the Apple hypervisor framework to run native binaries at native speed, I don't understand the performance issue.

But really, what is the alternative to run `containerd` on macOS?

**Reply by rfay** [reply]
The issue here is about architectures. Linux can be amd64 (traditional Intel) or arm64 (various machines including the Mac M1). 

QEMU is able to translate between architectures, the Apple Hypervisor framework does not do that at all. It has no capability of running amd64 images on the mac M1. Even parallels can't run amd64 images. Rosetta does an admirable job with binary executables. 

QEMU's capabilities are awesome, but the costs of doing it are significant, generally considered to be a 6x slowdown. And the reliability issues are always there when running cross-architecture emulation. They're different on each docker image. 

So the two things to run native are:
* Native binaries (not absolutely required, but easy to generate and solve various problems). Easy to build cross-architecture with golang.
* Native docker images, so that no qemu emulation is required.

**Reply by rfay** [reply]
Oh, I see you *do* have arm64 binaries. But using qemu to run amd64 containers when on arm64 architecture, right?

**Reply by afbjorklund** [reply]
There are two different variants of QEMU, one is called "[qemu-user](https://www.qemu.org/docs/master/user/)" and is used to run individual programs from a different architecture. This is mapped into binfmt_misc, and allows Linux to run foreign architecture (similar to Apple's Rosetta). The other one is called "[qemu-system](https://www.qemu.org/docs/master/system/)", and it runs the entire operating system (similar to Oracle VirtualBox). Including drivers and such, where the actual CPU execution can be done in hardware (virtualization) or in software (emulation, "[TCG](https://wiki.qemu.org/Documentation/TCG)").

If you run virtualization on a modern hardware with acceleration, you expect to see at most 10% slowdown (more like 5%, it varies). But if you emulate another processor, you expect to see more of a 10x slowdown (can be a lot more). So in both cases, you want to run binaries from the same architecture. In this context, we are mostly concerned about two architectures: x86_64/amd64 and aarch64/arm64, out of the 32 or so architectures that QEMU includes out-of-the-box.

Anyway, during normal execution of Lima with a multi-arch manifest, there is **no** emulation being performed at all.

You would still get better performance by running on bare metal, but that also requires that you are running Linux...

**Reply by jandubois** [reply]
> Oh, I see you _do_ have arm64 binaries. But using qemu to run amd64 containers when on arm64 architecture, right?

No. We are using qemu on arm64 to run a Linux VM using the native arm64 instruction set. It is used for the hypervisor capabilities, not for the emulation ones. Because the container runtime only runs on Linux. But we try to run the guest OS with the same instruction set as the host, so we don't need emulation.

Now, Lima does support running an amd64 version of Linux on arm64, but it is incredibly slow. As @afbjorklund explained above, you can use `qemu-user-amd64` with `binfmt_misc` inside the arm64 Linux VM to just emulate the container workload and not the whole OS. It is still a lot slower than running arm64 containers though because there is no hardware acceleration in this case.

The choice is really up to the user.

**Reply by afbjorklund** [reply]
It does seem like the cross-platform support (for containers) is not included by default, maybe it should be ?

`lima nerdctl run arm64v8/alpine arch`

`lima nerdctl run amd64/alpine arch`

The VM needs to include `qemu-user-static` (and `binfmt-support`), or else you will get the dreaded message:

```
standard_init_linux.go:228: exec user process caused: exec format error
```

`lima sudo apt install qemu-user-static`

**Reply by jandubois** [reply]
> The VM needs to include `qemu-user-static` (and `binfmt-support`), or else you will get the dreaded message:

The `alpine-lima` image does include `qemu-aarch64` and configures it, so it should work automatically. But I just noticed that the `aarch64` version of the iso should bundle `qemu-x86_64` instead, but doesn't.

**Reply by afbjorklund** [reply]
@jandubois : it also seems to only have the dynamic version, which relies on the libc _in the container_

```
lima-alpine:~$ ldd /usr/bin/qemu-aarch64
	/lib/ld-musl-x86_64.so.1 (0x7fccd816b000)
```

So you need to install the static versions, i.e. `qemu-aarch64-static` and `qemu-x86_64-static`

**Reply by jandubois** [reply]
> So you need to install the static versions, i.e. `qemu-aarch64-static` and `qemu-x86_64-static`

Do you happen to know where to get them?

`qemu-aarch64-static` does not seem to exist in the default repos.

Also, can you show an example container that fails with `qemu-aarch64` but would succeed with `qemu-aarch64-static`?

**Reply by AkihiroSuda** [reply]
`tonistiigi/binfmt` image contains static qemu-user with several important patches: https://github.com/tonistiigi/binfmt

I'm planning to add the qemu-user binaries from `tonistiigi/binfmt` to `nerdctl-full-*.tar.gz` (which is installed into Lima instances).
https://github.com/containerd/nerdctl/issues/412 .

**Reply by afbjorklund** [reply]
> Do you happen to know where to get them?

Those were the names on Ubuntu, sorry.

**Reply by jandubois** [reply]
> `tonistiigi/binfmt` image contains static qemu-user with several important patches: https://github.com/tonistiigi/binfmt

I'm confused how this actually works (I know it does work, but don't understand how): the emulators only exist inside the container, so how can they be invoked from the host:

```console
root@lima-default:~# cat /proc/sys/fs/binfmt_misc/qemu-aarch64
enabled
interpreter /usr/bin/qemu-aarch64
flags: OCF
offset 0
magic 7f454c460201010000000000000000000200b700
mask ffffffffffffff00fffffffffffffffffeffffff
root@lima-default:~# ls -l /usr/bin/qemu*
ls: cannot access '/usr/bin/qemu*': No such file or directory
root@lima-default:~# docker run --platform aarch64 busybox uname -a
Linux 94c727fedb8e 5.13.0-14-generic #14-Ubuntu SMP Mon Aug 2 12:43:35 UTC 2021 aarch64 GNU/Linux
```

Is the interpreter somehow copied during registration, or what is the magic that makes this work?

As to the "important patches", aren't those the patches from Alpine, so shouldn't the Alpine packages already have them?

**Reply by jandubois** [reply]
> what is the magic that makes this work?

Ok, I looked it up myself: It is the `F` flag from `flags: OCF`:

>The usual behaviour of binfmt_misc is to spawn the binary lazily when the misc format file is invoked. However, this doesn’t work very well in the face of mount namespaces and changeroots, so the F mode opens the binary as soon as the emulation is installed and uses the opened image to spawn the emulator, meaning it is always available once installed, regardless of how the environment changes.

**Reply by AkihiroSuda** [reply]
> Is the interpreter somehow copied during registration, or what is the magic that makes this work?

The magic is the "F" flag (`fix binary`)
https://www.kernel.org/doc/html/latest/admin-guide/binfmt-misc.html

> As to the "important patches", aren't those the patches from Alpine, so shouldn't the Alpine packages already have them?

Doesn't seem so.
https://github.com/tonistiigi/binfmt/tree/master/patches
https://git.alpinelinux.org/aports/tree/community/qemu?h=master

**Reply by rfay** [reply]
Thanks all for the patient explanations! I look forward to getting more familiar with the details. I guess I had only been around `qemu-system`.

**Reply by afbjorklund** [reply]
@jandubois : on ubuntu, they are installed with the `binfmt-support` package (a dependency of `qemu-user-static`)

**Reply by afbjorklund** [reply]
@rfay : yet all your comments were about `qemu-user` ? Anyway, lima should support both of them - i.e. if you need it

**Reply by rfay** [reply]
No, really the comments were a worry that this was using qemu-system to do cross-architecture runs, but from all the kind responses here I guess that's not the case. I look forward to getting more familiar with this system.

**Reply by afbjorklund** [reply]
The virtual machine will have to run with qemu-system, but hopefully it will be accelerated (hvf/kvm)...
The containers don't have to run qemu-user, if they have right architecture.

**Reply by rfay** [reply]
My original worry was that the VM was set up as amd64 and would have to run cross-architecture on an arm64 machine.

**Comment by afbjorklund** [comment]
> But really, what is the alternative to run containerd on macOS?

The alternative would be jails (from FreeBSD), but they are not there.

Then again, WSL 1 ran containers (on Windows). Look how far it got ?

So I guess your best bet is to run a **real** native binary, i.e. `darwin`
